### PR TITLE
Refactor vLLM + Embed support

### PR DESCRIPTION
**Pull request by gavrissh**
**What this PR does / why we need it**:

- Get Embedding into vllm
- Refactor VLLM model to keep better maintainability
- support engine in model class

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
Add Support for OpenAI-compatible Embeddings API in HuggingFace Serving Runtime for vLLM integration
```

**Re-running failed tests**

- `/rerun-all` - rerun all failed workflows.
- `/rerun-workflow <workflow name>` - rerun a specific failed workflow. Only one workflow name can be specified. Multiple /rerun-workflow commands are allowed per comment.

### CODE REVIEW

**Comment by gavrissh** [comment]
Dev Code is ready, Completing some final testing

File: python/kserve/kserve/protocol/rest/openai/openai_chat_adapter_model.py [hunk 1]
```diff
@@ -197,43 +196,58 @@ def completion_to_chat_completion_chunk(
...
-            completion_iterator = cast(
-                AsyncIterator[Completion],
-                await self.create_completion(completion_request),
+            completion_iterator = await self.create_completion(
+                completion_params, raw_request
             )
 
-            def mapper(completion: Completion) -> ChatCompletionChunk:
+            # Since vllm must support Python 3.8, we can't use str.removeprefix(prefix)
+            # introduced in Python 3.9
```
Comment (sivanantha321) [inline python/kserve/kserve/protocol/rest/openai/openai_chat_adapter_model.py]: we don't support python 3.8

File: python/kserve/kserve/protocol/rest/openai/types/__init__.py [hunk 2]
```diff
@@ -12,72 +12,82 @@
...
+    CompletionResponseStreamChoice as CompletionChunkChoice,
+    CompletionLogProbs,
+    UsageInfo,
+    ChatCompletionLogProbsContent,
+    ModelCard as Model,
+    ModelList,
 )
-from kserve.protocol.rest.openai.types.openapi import (
-    CreateChatCompletionStreamResponse as ChatCompletionChunk,
+from vllm.entrypoints.openai.protocol import ChatCompletionRequest, ChatCompletionResponse as ChatCompletion
```
Comment (sivanantha321) [inline python/kserve/kserve/protocol/rest/openai/types/__init__.py:32]: Are we going to make vLLM as dependency for KServe ?

Comment (yuzisun) [reply python/kserve/kserve/protocol/rest/openai/types/__init__.py:32]: Does vllm extend openai protocol for additional feature support?

Comment (gavrissh) [reply python/kserve/kserve/protocol/rest/openai/types/__init__.py:32]: Yes vllm already supports additional params over OpenAI Object spec

Comment (gavrissh) [reply python/kserve/kserve/protocol/rest/openai/types/__init__.py:32]: >Are we going to make vLLM as dependency for KServe ?

@sivanantha321 Yes, had proposed this approach in the previous KServe Meeting

File: python/kserve/kserve/protocol/rest/openai/openai_chat_adapter_model.py (hunk 1 above)
Comment (gavrissh) [reply python/kserve/kserve/protocol/rest/openai/openai_chat_adapter_model.py]: removed

File: .github/workflows/artexplainer-docker-publish.yml [hunk 3]
```diff
@@ -39,7 +39,7 @@ jobs:
       - name: Run tests
         uses: docker/build-push-action@v5
         with:
-          platforms: linux/amd64, linux/arm64/v8
+          platforms: linux/amd64
```
Comment (gavrissh) [inline .github/workflows/artexplainer-docker-publish.yml]: This change was made to enable running live huggingfaceserver with vLLM for testing. (It was failing on `linux/arm64/v8`)

Let me know if this is a mandatory platform. As I did see it was not used in all the docker tests

File: python/huggingfaceserver/huggingfaceserver/encoder_model.py [hunk 4]
```diff
@@ -353,67 +357,85 @@ def _log_request(self, request_id: str, prompt: list[str]) -> None:
...
+                            err_type="invalid_request_error",
+                        )
+                    )
+
+            # Call the inference to determine the embedding values
+            context = {}
+            instances = (
+                request.input if isinstance(request.input, list) else [request.input]
+            )
+            inference_out, _ = await self({"instances": instances}, context)
```
Comment (Datta0) [inline python/huggingfaceserver/huggingfaceserver/encoder_model.py:404]: curious what the `context` is supposed to be... I see we're passing empty dict and trying to modify it in-place. Isn't returning the value a better approach?

File: python/huggingfaceserver/huggingfaceserver/generative_model.py [hunk 5]
```diff
@@ -180,14 +201,15 @@ def __init__(
...
-            raise RuntimeError(
+            raise OpenAIError(
                 f"Generative model does not support encoder-only task: {self.task.name}"
             )
 
     def load(self) -> bool:
         model_id_or_path = self.model_id_or_path
 
         self.max_length = _get_and_verify_max_len(self.model_config, self.max_length)
+        print(self.max_length)
```
Comment (Datta0) [inline python/huggingfaceserver/huggingfaceserver/generative_model.py]: Please remove. Also check other occurrences and sanitise the file

File: python/kserve/kserve/model.py [hunk 6]
```diff
@@ -79,10 +80,18 @@ def start(self):
         """Start handler can be overridden to perform model setup"""
         self.ready = True
 
+    async def start_engine(self):
+        """Certain models may require an engine to be started before they can be used"""
+        self.ready = True
```
Comment (Datta0) [inline python/kserve/kserve/model.py:85]: Thought: Should we make this abstract method then? Force consumer to implement?

File: python/huggingfaceserver/tests/setup_vllm.sh [hunk 7]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+VLLM_VERSION=v0.6.6.post1
+VLLM_DIR=vllm-clone
+
+mkdir $VLLM_DIR
+cd $VLLM_DIR
+git clone --branch $VLLM_VERSION --depth 1 https://github.com/vllm-project/vllm.git . && \
```
Comment (Datta0) [inline python/huggingfaceserver/tests/setup_vllm.sh]: Can you explain why we're building vLLM instead of doing pip install here?

Comment (gavrissh) [reply python/huggingfaceserver/tests/setup_vllm.sh]: This is for setting up vLLM+openvino for tests

File: python/kserve/kserve/model.py (hunk 6 above)
Comment (gavrissh) [reply python/kserve/kserve/model.py:85]: Followed similar pattern as other methods

File: python/huggingfaceserver/tests/test_vllm_generative.py [hunk 8]
```diff
@@ -0,0 +1,1610 @@
...
+MODEL_NAME = "test-model"
+GUIDED_DECODING_BACKENDS = ["outlines", "lm-format-enforcer", "xgrammar"]
+
+
+@pytest.fixture(scope="module")
+def server():  # noqa: F811
+    args = [
+        # use half precision for speed and memory savings in CI environment
+        "--dtype",
+        "bfloat16",
```
Comment (Datta0) [inline python/huggingfaceserver/tests/test_vllm_generative.py]: I'd suggest using float16. It is more widely supported and thats the default for our inference.

**Comment by gavrissh** [comment]
@yuzisun @sivanantha321 This is ready for review!

**Review by Datta0 — APPROVED** [review]

File: python/kserve/kserve/protocol/rest/openai/openai_model.py [hunk 9]
```diff
@@ -80,28 +61,38 @@ class OpenAICompletionModel(OpenAIModel):
...
+        self,
+        request: ChatCompletionRequest,
+        raw_request: Optional[Request] = None,
+        context: Optional[Dict[str, Any]] = None,
+    ) -> Union[AsyncGenerator[str, None], ChatCompletion, ErrorResponse]:
         pass
 
 
-class OpenAIEmbeddingModel(OpenAIModel):
+class OpenAIEncoderModel(OpenAIModel):
```
Comment (yuzisun) [inline python/kserve/kserve/protocol/rest/openai/openai_model.py:81]: @gavrissh @Datta0 is the goal here to support other encoder models here?
vllm supports embed,classify,reward,score and interestingly vllm can do fallback now to transformers.

Comment (yuzisun) [reply python/kserve/kserve/protocol/rest/openai/openai_model.py:81]: encode returns the hidden state directly and vllm has different APIs for embed, classify and score.
https://docs.vllm.ai/en/v0.6.5/models/pooling_models.html

Comment (Datta0) [reply python/kserve/kserve/protocol/rest/openai/openai_model.py:81]: > @gavrissh @Datta0 is the goal here to support other encoder models here? vllm supports embed,classify,reward,score and interestingly vllm can do fallback now to transformers.

Yeah the idea is to support as wide range of models as vLLM supports and we can possibly do via kserve.
I have to check how extensively the transformers 4.48 and vLLM fallback is...

**Comment by yuzisun** [comment]
@gavrissh could we add a test for tool configuration to verify the change?

**Comment by yuzisun** [comment]
> @gavrissh could we add a test for tool configuration to verify the change?

https://platform.openai.com/docs/guides/function-calling

**Comment by gavrissh** [comment]
> @gavrissh could we add a test for tool configuration to verify the change?

Sure, will add it!

**Comment by gavrissh** [comment]
> > @gavrissh could we add a test for tool configuration to verify the change?
> 
> Sure, will add it!

I have added the tool configuration test

**Comment by gavrissh** [comment]
@yuzisun This is ready, could you suggest if there are any other items to include

**Comment by yuzisun** [comment]
> @yuzisun This is ready, could you suggest if there are any other items to include

Thanks! sorry need you to help rebase after merging https://github.com/kserve/kserve/pull/4250

**Comment by gavrissh** [comment]
> 
> Thanks! sorry need you to help rebase after merging #4250

I have resolved the conflicts and all tests have passed

File: python/huggingfaceserver/huggingfaceserver/encoder_model.py [hunk 10]
```diff
@@ -278,7 +282,7 @@ async def predict(
                         outputs = outputs.logits
                     return outputs
             except Exception as e:
-                raise InferenceError(str(e))
+                raise OpenAIError(str(e))
```
Comment (sivanantha321) [inline python/huggingfaceserver/huggingfaceserver/encoder_model.py]:
```suggestion
                raise OpenAIError from e
```

File: python/huggingfaceserver/huggingfaceserver/encoder_model.py [hunk 11]
```diff
@@ -353,67 +357,86 @@ def _log_request(self, request_id: str, prompt: list[str]) -> None:
...
+                object="embedding",
+                data=data,
+                usage=UsageInfo(
+                    prompt_tokens=stats.num_prompt_tokens,
+                    total_tokens=stats.num_prompt_tokens,
+                ),
+            )
+
+        except Exception as e:
+            print(e)
```
Comment (sivanantha321) [inline python/huggingfaceserver/huggingfaceserver/encoder_model.py]:
```suggestion
```

File: python/huggingfaceserver/huggingfaceserver/encoder_model.py [hunk 12]
```diff
@@ -353,67 +357,86 @@ def _log_request(self, request_id: str, prompt: list[str]) -> None:
...
+            raise OpenAIError(f"Error during embedding creation: {str(e)}")
```
Comment (sivanantha321) [inline python/huggingfaceserver/huggingfaceserver/encoder_model.py]:
```suggestion
            raise OpenAIError(f"Error during embedding creation: {e}") from e
```

File: python/huggingfaceserver/tests/setup_vllm.sh [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+VLLM_VERSION=v0.7.2
+VLLM_DIR=vllm-clone
+
+source $(poetry env info -p)/bin/activate
+
+mkdir $VLLM_DIR
+cd $VLLM_DIR
+git clone --branch $VLLM_VERSION --depth 1 https://github.com/vllm-project/vllm.git .
+pip install --upgrade pip && \
+sed -i 's/@main//' ./requirements-openvino.txt && \
```
Comment (sivanantha321) [inline python/huggingfaceserver/tests/setup_vllm.sh]: This is no longer needed.

File: python/kserve/kserve/protocol/rest/openai/endpoints.py [hunk 14]
```diff
@@ -80,24 +81,23 @@ async def create_completion(
...
-        if isinstance(completion, AsyncIterable):
-
-            async def stream_results() -> AsyncGenerator[str, None]:
-                async for partial_completion in completion:
-                    yield f"data: {partial_completion.model_dump_json()}\n\n"
-                yield "data: [DONE]\n\n"
-
-            return StreamingResponse(stream_results(), media_type="text/event-stream")
+        if isinstance(completion, ErrorResponse):
+            return JSONResponse(
```
Comment (sivanantha321) [inline python/kserve/kserve/protocol/rest/openai/endpoints.py]: Any reason not to use ORJSONResponse ?

File: python/kserve/kserve/protocol/rest/openai/endpoints.py [hunk 15]
```diff
@@ -125,33 +125,30 @@ async def create_chat_completion(
...
-        if isinstance(completion, AsyncIterable):
-
-            async def stream_results() -> AsyncGenerator[str, None]:
-                async for chunk in completion:
-                    yield f"data: {chunk.model_dump_json()}\n\n"
-                yield "data: [DONE]\n\n"
-
-            return StreamingResponse(stream_results(), media_type="text/event-stream")
+        if isinstance(completion, ErrorResponse):
+            return JSONResponse(
```
Comment (sivanantha321) [inline python/kserve/kserve/protocol/rest/openai/endpoints.py]: Same as above

File: python/kserve/kserve/protocol/rest/openai/endpoints.py [hunk 16]
```diff
@@ -166,27 +163,35 @@ async def create_embedding(
...
             model_name=model_name,
-            request=request_body,
-            headers=request_headers,
+            request=params,
+            raw_request=raw_request,
+            headers=raw_request.headers,
             response=response,
         )
+        if isinstance(embedding, ErrorResponse):
+            return JSONResponse(
```
Comment (sivanantha321) [inline python/kserve/kserve/protocol/rest/openai/endpoints.py]: Same as above

**Comment by sivanantha321** [comment]
Everything looks good other than some small nits :).

File: python/kserve/kserve/protocol/rest/openai/endpoints.py (hunk 14 above)
Comment (gavrissh) [reply python/kserve/kserve/protocol/rest/openai/endpoints.py]: No particular reason, is it expected to use `ORJSONResponse`?

Comment (gavrissh) [reply python/kserve/kserve/protocol/rest/openai/endpoints.py]: I have updated

**Comment by sivanantha321** [comment]
Awesome work!

**Comment by gavrissh** [comment]
We can merge this now

**Comment by yuzisun** [comment]
/approve

Great work! thanks @gavrissh !!
